From 89350e8e9225a457cf1722905f8079b488d2ec91 Mon Sep 17 00:00:00 2001
From: Jesper Brouer <brouer@redhat.com>
Date: Wed, 19 Mar 2014 11:29:59 -0400
Subject: [net] ipv6: ipv6_find_hdr restore prev functionality

Message-id: <20140319112959.13556.58021.stgit@dragon>
Patchwork-id: 78405
O-Subject: [RHEL7 net PATCH 2/2] ipv6: ipv6_find_hdr restore prev functionality
Bugzilla: 1074516
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Thomas Graf <tgraf@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Bugzilla: 1074516
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7224056

Backport of upstream commit: accfe0e35632

 commit accfe0e356327da5bd53da8852b93fc22de9b5fc
 Author: Hans Schillstrom <hans@schillstrom.com>
 Date:   Thu Feb 27 12:57:58 2014 +0100

    ipv6: ipv6_find_hdr restore prev functionality

    The commit 9195bb8e381d81d5a315f911904cdf0cfcc919b8 ("ipv6: improve
    ipv6_find_hdr() to skip empty routing headers") broke ipv6_find_hdr().

    When a target is specified like IPPROTO_ICMPV6 ipv6_find_hdr()
    returns -ENOENT when it's found, not the header as expected.

    A part of IPVS is broken and possible also nft_exthdr_eval().
    When target is -1 which it is most cases, it works.

    This patch exits the do while loop if the specific header is found
    so the nexthdr could be returned as expected.

    Reported-by: Art -kwaak- van Breemen <ard@telegraafnet.nl>
    Signed-off-by: Hans Schillstrom <hans@schillstrom.com>
    CC:Ansis Atteka <aatteka@nicira.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jesper Dangaard Brouer <brouer@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/exthdrs_core.c b/net/ipv6/exthdrs_core.c
index 140748d..8af3eb5 100644
--- a/net/ipv6/exthdrs_core.c
+++ b/net/ipv6/exthdrs_core.c
@@ -212,7 +212,7 @@ int ipv6_find_hdr(const struct sk_buff *skb, unsigned int *offset,
   found = (nexthdr == target);
 
   if ((!ipv6_ext_hdr(nexthdr)) || nexthdr == NEXTHDR_NONE) {
-   if (target < 0)
+   if (target < 0 || found)
     break;
    return -ENOENT;
   }
-- 
1.7.1